### ISSUE TITLE
Create wheels with bundled SDL.dll

### ISSUE THREAD

**Issue by marcusva** [marcusva/py-sdl2#78]
Originally reported by: **anatoly techtonik (Bitbucket: [techtonik](https://bitbucket.org/techtonik), GitHub: [techtonik](https://github.com/techtonik))**

----------------------------------------

It will help to make `py-sdl2` more usable.

----------------------------------------
- Bitbucket: https://bitbucket.org/marcusva/py-sdl2/issue/78

**Comment by a-hurst** [comment]
Have any efforts been made on this? Installation on linux and macOS is pretty straightforward, but the lack of a simple .exe installer for SDL2 on Windows makes it much harder to officially support the platform for any project that uses PySDL2. Having the core and peripheral (ttf, mixer, etc.) libraries included with PySDL2 would definitely make life easier there.

I know some Python packages (e.g. PyAudio) only bother including the required libraries on Windows because it's only really painful there, so doing the same and keeping this Windows-only (at least at first) might make the issue easier to tackle.

**Comment by marcusva** [comment]
No, there has not been done any work on that issue so far.

**Comment by a-hurst** [comment]
Okay, thanks for letting me know! I started looking into how other Python packages do this, and the solutions range from clean to horrifying. The best one I found was [SoundFile](https://github.com/bastibe/SoundFile) (a CFFI wrapper for libsndfile), which keeps the compiled libraries in a git submodule so they don't bloat up the main repository, and essentially just grafts the correct dll to the package during install/wheel building.

As a proof-of-concept, I spent a bit of time today compiling all the SDL2 dlls for my OS (macOS) and adapting pysdl2's setup.py and dll.py based on the SoundFile code to a) include them on install and b) load them at runtime, with partial success (it works, but only when the libraries sdl2_ttf/sdl2_mixer/sdl2_image depend on are already installed, so I need to build those statically into the SDL2 libs). I'll keep you posted on my progress.

A question regarding implementation: if SDL2 libraries are being included in the package, should they always take precedence over system-installed SDL2 and related libraries? If so, should there be a flag or environment variable users can set to ignore them and use ones included in their application directory or system?

**Comment by marcusva** [comment]
If PySDL2 ships with its own set of DLLs for certain platforms, I'd want those to be the default choice.

Since the bundled DLLs are those that were tested against and user- or system-specific DLLs may behave differently, using the shipped ones as default hopefully ensures less issue reports.
However, an environment flag should override that behaviour by checking the system first, before falling back to the shipped ones.

Some minor points to consider:

1. 32-bit and 64-bit DLL variants and builds will be necessary and need to be considered
2. I'd want to stick to the official SDL2 DLL binaries, wherever possible, so that issues with SDL2 itself won't be caused by the PySDL2 build environment
3. Which platforms should be supported by default? Windows and MacOS X are supported with pre-compiled packages by SDL2 upstream, which should be sufficient in my opinion. Most Linux and BSD systems come with their own build tools and pre-compiled packages, so those do not need to be supported.
4. SDL_gfx - there are no pre-compiled binaries provided by upstream. Is there a way to avoid additional maintenance efforts to build and/or rebuild it with newer SDL2 releases?
5. Maintenance and downwards compatibility. Should the DLLs be updated on each PySDL2 release or be updated depending on the supported feature sets of PySDL2?

The last point is not of huge importance, since the current PySDL2 release cycles are currently really, really long :-)

**Comment by a-hurst** [comment]
For points 1 through 3: the official pre-built SDL2 frameworks for macOS are both 32-bit and 64-bit, and work quite nicely with pysdl2 when included into the package itself (with a little editing in the dll.py file). The official Windows 64-bit DLLs also seem to work find when bundled in the package (I don't have a 32-bit Windows VM available, but I'd assume it'd work too).

The only challenge now is to figure out how to properly include the correct DLLs for each platform with the package, while excluding the DLLs for other platforms/architectures: I haven't spent a ton of time on this since I got side-tracked working on documentation for another project, but I'll take another crack at it this week. I'll see how much of the SoundFile setup.py code can be adapted for this.

The other problem, like you said, is the lack of official pre-built SDL_gfx binaries for either platform. The options I can think of here are:

a) Leave this as an external dependency users will have to install themselves if they want it, or
b) Come up with a build script that builds DLLs for both Win32/64 and macOS, and hope it doesn't get updated too often. If we're interested in building a universal 32/64-bit framework for macOS in line with the other pre-built ones, I'll need to dig up an older Mac since official 32-bit compiling support was dropped long ago.

As for updating the DLLs (or not) with every release, I'll check how other Python packages that package large C libraries work. There may be a solution involving subpackages and dependencies that would prevent it from re-downloading the whole ~10MB of libraries with every update.

**Comment by a-hurst** [comment]
Okay, so almost a year later (this project fell by the wayside) I think I have a good solution to this! There's currently no support for SDL_gfx yet, and no Linux support whatsoever (though these could probably be added in later using build scripts with TravisCI & Appveyor), but I've written a Python package that fetches and installs the official pre-built SDL2 + SDL2_Mixer + SDL2_ttf + SDL2_image binaries for macOS, Windows 32-bit, or Windows 64-bit in a way such that PySDL2 can easily use them. Adding official support for this in PySDL2 would not involve much effort at all on your part.

Basically, the way I have it set up now, the package's setup.py script determines the current OS and (if supported) downloads the official SDL2 binaries for the platform and copies them into the package folder. The package itself is only a short \_\_init\_\_.py file, which checks if `PYSDL2_DLL_PATH` has been set yet and if it hasn't, sets the value of the environment variable to the folder with the bundled system-specific SDL2 DLLs within the package. Then, when `import sdl2` is run, it finds and loads those DLLs:

```python
import os

dll_path = os.getenv('PYSDL2_DLL_PATH')
if dll_path == None:
	root_path = os.path.abspath(os.path.dirname(__file__))
	dll_path = os.path.join(root_path, 'dll')
	os.environ['PYSDL2_DLL_PATH'] = dll_path
```

The main change required in PySDL2 itself to make integration as smooth as possible (other than a few small bugfixes to get the binaries working as intended) would be to add

```python
try:
    import sdl2dll
except (ModuleNotFoundError, ImportError):
    pass
```

at the top of dll.py. Then, if the sdl2dll package is installed, it'll set the DLL path to the correct location prior to attempting to load the DLLs, and will default to normal behaviour if it isn't. 

This has a few advantages over bundling the DLLs with PySDL2 itself:
* It means that you can update the PySDL2 wheels without making people re-download the whole ~30MB of SDL2 + extras every time
* It means that people can continue to use the system-installed versions of SDL2 if they want to
* It requires minimal modification of the existing code

You can check out what I've done [here](https://github.com/a-hurst/pysdl2-dll). Let me know what you think!

**Comment by a-hurst** [comment]
@marcusva Just as an update on this, yesterday I hacked together [an SDL2_gfx buildbot](https://github.com/a-hurst/sdl2gfx-builds) with Appveyor and TravisCI that produces working binaries for both Windows (32/64) and macOS (10.6+ and 64-only, same as the official SDL2 frameworks), meaning that SDL2_gfx support can be added to the pysdl2-dll package.

Apart from some extra testing, the next steps on my end now are figuring out how to build python-independent-but-os-dependent wheels for PyPi and figuring out how to bundle all the relevant licence info from the separate SDL2_* projects in the wheels.

Before I put too much more effort into it though, can I get your thoughts on this approach (i.e. having the official binaries in a separate package that, if not otherwise specified, sets the PYSDL_DLL_PATH variable to its own internal dll folder)? Just want to make sure you're on board before I spent too many more hours getting this working.
